### PR TITLE
Mark pop/clear error stack in der2key_decode_p8

### PR DESCRIPTION
**Pull request by danbev**
This commit sets the error mark before calling `d2i_X509_SIG`
and clear it if that function call is successful.

The motivation for this is that if `d2i_X509_SIG` returns NULL then the
else clause will be entered and `d2i_PKCS8_PRIV_KEY_INFO` will be
called. If `d2i_X509_SIG` raised any errors those error will be on the
error stack when `d2i_PKCS8_PRIV_KEY_INFO` gets called, and even if it
returns successfully those errors will still be on the error stack.

We ran into this issue when upgrading Node.js to 3.0.0-alpha15.
More details can be found in the ref links below.

Refs: https://github.com/nodejs/node/issues/38373
Refs: https://github.com/danbev/learning-libcrypto/blob/master/notes/wrong-tag-issue2.md

##### Checklist
- [x] tests are added or updated

### CODE REVIEW

File: providers/implementations/encode_decode/decode_der2key.c [hunk 1]
```diff
@@ -136,6 +138,8 @@ static void *der2key_decode_p8(const unsigned char **input_der,
             ctx->flag_fatal = 1;
         X509_SIG_free(p8);
     } else {
+        // Pop any errors that might have been raised by d2i_X509_SIG.
```
Comment (t8m) [inline providers/implementations/encode_decode/decode_der2key.c:141]: Use `/**/`

File: test/evp_extra_test.c [hunk 2]
```diff
@@ -1169,6 +1169,43 @@ static int test_EVP_PKCS82PKEY(void)
...
+
+    ok = 1;
+ done:
+    EVP_PKEY_free(pkey);
+    PKCS8_PRIV_KEY_INFO_free(p8inf);
+    BIO_free_all(membio);
+    OSSL_PROVIDER_unload(provider);
+    return ok;
+}
 #endif
```
Comment (t8m) [inline test/evp_extra_test.c:1209]: The `#endif` should be before the added test.

**Review by t8m — CHANGES_REQUESTED** [review]

File: providers/implementations/encode_decode/decode_der2key.c (hunk 1 above)
Comment (danbev) [reply providers/implementations/encode_decode/decode_der2key.c]: Sorry about that, I've updated now. Thanks

File: providers/implementations/encode_decode/decode_der2key.c [hunk 3]
```diff
@@ -126,9 +126,9 @@ static void *der2key_decode_p8(const unsigned char **input_der,
 
     ERR_set_mark();
     if ((p8 = d2i_X509_SIG(NULL, input_der, input_der_len)) != NULL) {
-        ERR_clear_last_mark();
         char pbuf[PEM_BUFSIZE];
         size_t plen = 0;
+        ERR_clear_last_mark();
```
Comment (t8m) [inline providers/implementations/encode_decode/decode_der2key.c:131]: Please add empty line after declarations.

File: test/evp_extra_test.c (hunk 2 above)
Comment (danbev) [reply test/evp_extra_test.c]: Updated now, thanks.

File: providers/implementations/encode_decode/decode_der2key.c (hunk 3 above)
Comment (danbev) [reply providers/implementations/encode_decode/decode_der2key.c:132]: Added now

**Review by levitte — APPROVED** [review]

**Comment by danbev** [comment]
I've tried to reproduce the failing [address sanitizer test](https://github.com/openssl/openssl/pull/15067/checks?check_run_id=2456722114) locally but I've not been able to reproduce these failures.
Has anyone run into something like this before?

File: test/evp_extra_test.c [hunk 4]
```diff
@@ -1169,7 +1169,46 @@ static int test_EVP_PKCS82PKEY(void)
 
     return ret;
 }
+
 #endif
+static int test_EVP_PKCS82PKEY_wrong_tag(void)
+{
+    OSSL_PROVIDER *provider = OSSL_PROVIDER_load(NULL, "default");
```
Comment (t8m) [inline test/evp_extra_test.c:1176]: Why do you load the provider here? It could be the source of the failing sanitizer test problem.

Comment (danbev) [reply test/evp_extra_test.c:1176]: I ran into an issue when running the test suite:
```console
$ make -j8 test
...
30-test_evp.t ...................... ok    
30-test_evp_extra.t ................ 1/? 
    # ERROR: (int) 'ERR_get_error() == 0' failed @ test/evp_extra_test.c:1199
    # [503841036] compared to [0]
    # OPENSSL_TEST_RAND_ORDER=1619614042
    not ok 8 - test_EVP_PKCS82PKEY_wrong_tag
# ------------------------------------------------------------------------------
../../util/wrap.pl ../../test/evp_extra_test -context => 1
not ok 2 - running evp_extra_test with a non-default library context
30-test_evp_extra.t ................ 3/? ---------------------------------------
30-test_evp_extra.t ................ Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/3 subtests
```
which I did not get when running the test standalone:
```console
$ ./test/evp_extra_test --test test_EVP_PKCS82PKEY_wrong_tag
ok 1 - test_EVP_PKCS82PKEY_wrong_tag
```
So I added the provider here to try to work around this issue and just went by the error message mentioning using a non-default library context.

Comment (t8m) [reply test/evp_extra_test.c:1176]: Just add at the beginning of the test case something like:
```
if (testctx != NULL)
    /* test not supported with non-default context */
    return 1;
```

File: test/evp_extra_test.c [hunk 5]
```diff
@@ -1169,7 +1169,46 @@ static int test_EVP_PKCS82PKEY(void)
...
+    if (!TEST_ptr(membio = BIO_new(BIO_s_mem()))
+        || !TEST_ptr(pkey = load_example_rsa_key())
+        || !TEST_int_gt(i2d_PKCS8PrivateKey_bio(membio, pkey, NULL,
+                                                NULL, 0, NULL, NULL),
+                        0)
+        || !TEST_int_gt(membuf_len = BIO_get_mem_data(membio, &membuf), 0)
+        || !TEST_ptr(membuf)
+        || !TEST_mem_eq(membuf, (size_t)membuf_len,
+                        kExampleRSAKeyPKCS8, sizeof(kExampleRSAKeyPKCS8))
+        || !TEST_int_gt(PEM_write_bio_PKCS8PrivateKey(membio, pkey, NULL,
```
Comment (t8m) [inline test/evp_extra_test.c:1194]: I am confused by this - why you're also writing a PEM data there?

File: test/evp_extra_test.c (hunk 4 above)
Comment (danbev) [reply test/evp_extra_test.c:1176]: Ah great, I'll add that. Thanks!

File: test/evp_extra_test.c (hunk 5 above)
Comment (danbev) [reply test/evp_extra_test.c:1194]: I meant to remove that as it was part of the test that I based this test on. I'll remove this and update shortly. Thanks

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@levitte still OK?

**Comment by t8m** [comment]
ping @levitte for re-review

**Review by kaduk — APPROVED** [review]

**Review by levitte — APPROVED** [review]

**Comment by kaduk** [comment]
Pushed to master, so closing.
Thanks for the submission and working to resolve the review comments!

**Pull request closed without merging**